spring-kafka/src/main/java/org/springframework/kafka/AsyncKafkaTemplate.java

```diff
@@ -0,0 +1,229 @@
+import org.springframework.scheduling.TaskScheduler;
+import org.springframework.util.Assert;
+import org.springframework.util.concurrent.ListenableFuture;
+import org.springframework.util.concurrent.SettableListenableFuture;
+
+/**
+ * Async send/receive template.
+ *
+ * @param <K> the key type.
+ * @param <V> the outbound data type.
+ * @param <R> the reply data type.
+ *
+ * @author Gary Russell
+ * @since 2.1.3
+ *
+ */
+public class AsyncKafkaTemplate<K, V, R> extends KafkaTemplate<K, V> implements BatchMessageListener<K, R>,
+		InitializingBean, SmartLifecycle {
+
+	private static final long DEFAULT_REPLY_TIMEOUT = 5000L;
+
+	private final GenericMessageListenerContainer<K, R> replyContainer;
+
+	private final ConcurrentMap<String, RequestReplyFuture<K, V, R>> futures = new ConcurrentHashMap<>();
+
+	private TaskScheduler scheduler;
+
+	private boolean running;
+
+	private int phase;
+
+	private boolean autoStartup;
+
+	private long replyTimeout = DEFAULT_REPLY_TIMEOUT;
+
+	public AsyncKafkaTemplate(ProducerFactory<K, V> producerFactory,
+			GenericMessageListenerContainer<K, R> replyContainer) {
+		this(producerFactory, replyContainer, false);
+	}
+
+	public AsyncKafkaTemplate(ProducerFactory<K, V> producerFactory,
+			GenericMessageListenerContainer<K, R> replyContainer, boolean autoFlush) {
+		super(producerFactory, autoFlush);
+		this.replyContainer = replyContainer;
+		this.replyContainer.setupMessageListener(this);
+	}
+
+	protected void setTaskScheduler(TaskScheduler scheduler) {
+		this.scheduler = scheduler;
+	}
+
+	protected void setReplyTimeout(long replyTimeout) {
+		Assert.isTrue(replyTimeout >= 0, "'replyTimeout' must be >= 0");
+		this.replyTimeout = replyTimeout;
+	}
+
+	@Override
+	public void afterPropertiesSet() throws Exception {
+		Assert.notNull(this.scheduler, "A scheduler is required for timeouts");
+	}
+
+	@Override
+	public synchronized void start() {
+		try {
+			afterPropertiesSet();
+		}
+		catch (Exception e) {
+			throw new KafkaException("Failed to initialize", e);
+		}
+		this.replyContainer.start();
+		this.running = true;
+	}
+
+	@Override
+	public synchronized void stop() {
+		this.running = false;
+		this.replyContainer.stop();
+	}
+
+	@Override
+	public synchronized boolean isRunning() {
+		return this.running;
+	}
+
+	@Override
+	public int getPhase() {
+		return this.phase;
+	}
+
+	public void setPhase(int phase) {
+		this.phase = phase;
+	}
+
+	@Override
+	public boolean isAutoStartup() {
+		return this.autoStartup;
+	}
+
+	public void setAutoStartup(boolean autoStartup) {
+		this.autoStartup = autoStartup;
+	}
+
+	@Override
+	public void stop(Runnable callback) {
+		stop();
+		callback.run();
+	}
+
+	public RequestReplyFuture<K, V, R> sendAndReceive(ProducerRecord<K, V> record) {
+		String correlationId = createCorrelationId();
+		record.headers()
+				.add(new RecordHeader(KafkaHeaders.CORRELATION_ID, correlationId.getBytes(StandardCharsets.UTF_8)));
```

Comment:
Looks like we need exactly `byte[]` for the value, so, let's make that `createCorrelationId()` to return `byte[]` for us!

The default one indeed can be what you do with UUID, but as this:
```
 ByteBuffer bb = ByteBuffer.wrap(new byte[16]);
 bb.putLong(uuid.getMostSignificantBits());
 bb.putLong(uuid.getLeastSignificantBits());

return bb.array();
```

This way we will be free of the `String` to `byte[]` conversion here.

WDYT?